File: frontend/src/services/personalization.ts
import { useState, useEffect } from 'react';

interface PersonalizationData {
  theme?: string;
  highlightKeywords?: string[];
  // Add other personalization fields as needed
}

export const fetchPersonalizationData = async (userId: string): Promise<PersonalizationData> => {
  // In a real application, this would make an API call to the backend's personalization endpoint
  console.log(`[Frontend Service] Fetching personalization data for user: ${userId}`);
  // Placeholder data for demonstration
  if (userId === "personalizedUser") {
    return {
      theme: "dark",
      highlightKeywords: ["Robotics", "AI Ethics"],
    };
  } else if (userId === "adminUser") {
      return {
          theme: "light",
          highlightKeywords: ["Deployment", "FastAPI"],
      };
  }
  return { theme: "light", highlightKeywords: [] };
};

export const applyPersonalizationToContent = (
  content: string,
  personalization: PersonalizationData
): string => {
  let processedContent = content;

  if (personalization.highlightKeywords && personalization.highlightKeywords.length > 0) {
    personalization.highlightKeywords.forEach(keyword => {
      // Using a more robust regex for whole word matching and case-insensitivity
      const regex = new RegExp(`\\b(${keyword})\\b`, 'gi');
      processedContent = processedContent.replace(regex, `<span style="background-color: #ffd700; font-weight: bold;">$1</span>`);
    });
  }
  // Add more content personalization logic here if needed (e.g., reordering sections, filtering)
  return processedContent;
};

export const usePersonalizedContent = (content: string, userId: string): {
  displayContent: string;
  theme: string;
  personalizationLoading: boolean;
} => {
  const [personalizationData, setPersonalizationData] = useState<PersonalizationData | null>(null);
  const [displayContent, setDisplayContent] = useState<string>(content);
  const [personalizationLoading, setPersonalizationLoading] = useState<boolean>(true);

  useEffect(() => {
    setPersonalizationLoading(true);
    fetchPersonalizationData(userId)
      .then(data => {
        setPersonalizationData(data);
        setDisplayContent(applyPersonalizationToContent(content, data));
      })
      .finally(() => setPersonalizationLoading(false));
  }, [content, userId]);

  const theme = personalizationData?.theme || "light";

  return { displayContent, theme, personalizationLoading };
};

File: backend/src/services/personalization.py
import re
from typing import Dict, Any

def get_user_personalization_data(user_id: str) -> Dict[str, Any]:
    """Fetches personalization data for a given user ID from a placeholder store.
    In a real application, this would query a database or cache for user preferences,
    learning history, or other personalization factors.
    """
    print(f"[Backend Service] Fetching personalization data for user: {user_id}")

    # Placeholder logic for demonstration
    if user_id == "personalizedUser":
        return {
            "theme": "dark",
            "highlight_keywords": ["Robotics", "AI Ethics"],
            "recommended_chapters": ["15-hri.md", "16-ethics-future-ai.md"]
        }
    elif user_id == "adminUser":
        return {
            "theme": "light",
            "highlight_keywords": ["Deployment", "FastAPI", "Qdrant"],
            "recommended_chapters": ["10-gazebo-sim.md", "11-isaac-sim.md"]
        }
    # Default personalization for other users
    return {
        "theme": "light",
        "highlight_keywords": [],
        "recommended_chapters": []
    }

def apply_personalization_to_chapter_content(content: str, personalization_data: Dict[str, Any]) -> str:
    """Applies personalization logic to chapter content.
    This could involve highlighting keywords, reordering sections, inserting custom notes, etc.
    """
    processed_content = content

    # Example: Highlight keywords
    highlight_keywords = personalization_data.get("highlight_keywords", [])
    for keyword in highlight_keywords:
        regex = r"\b(" + re.escape(keyword) + r")\b" # Use re.escape for safety
        processed_content = re.sub(regex, r"<span style=\"background-color: #ffd700; font-weight: bold;\">\1</span>", processed_content, flags=re.IGNORECASE)

    return processed_content
